src/core/huff_coder.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2025, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/huff_coder.h"
+
+#include "base/logging.h"
+
+extern "C" {
+#include "huff/huf.h"
+}
+
+using namespace std;
+
+namespace dfly {
+
+constexpr size_t kWspSize = HUF_CTABLE_WORKSPACE_SIZE;
+
+bool HuffmanEncoder::Load(std::string_view binary_data, std::string* error_msg) {
+  CHECK(!huf_ctable_);
+
+  huf_ctable_.reset(new HUF_CElt[HUF_CTABLE_SIZE_ST(255)]);
+  table_max_symbol_ = 255;
+
+  unsigned has_zero_weights = 0;
+  size_t read_size = HUF_readCTable(huf_ctable_.get(), &table_max_symbol_, binary_data.data(),
+                                    binary_data.size(), &has_zero_weights);
+
+  if (HUF_isError(read_size)) {
+    huf_ctable_.reset();
+    *error_msg = HUF_getErrorName(read_size);
+    return false;
+  }
+  if (read_size != binary_data.size()) {
+    *error_msg = "Corrupted data";
+    huf_ctable_.reset();
+    return false;
+  }
+
+  return true;
+}
+
+bool HuffmanEncoder::Build(const unsigned hist[], unsigned max_symbol, std::string* error_msg) {
+  CHECK(!huf_ctable_);
+  huf_ctable_.reset(new HUF_CElt[HUF_CTABLE_SIZE_ST(max_symbol)]);
+
+  unique_ptr<uint32_t[]> wrkspace(new uint32_t[HUF_CTABLE_WORKSPACE_SIZE_U32]);
+
+  size_t num_bits =
+      HUF_buildCTable_wksp(huf_ctable_.get(), hist, max_symbol, 0, wrkspace.get(), kWspSize);
+  if (HUF_isError(num_bits)) {
+    *error_msg = HUF_getErrorName(num_bits);
+    huf_ctable_.reset();
+    return false;
+  }
+  num_bits_ = static_cast<uint8_t>(num_bits);
+  table_max_symbol_ = max_symbol;
+  return true;
+}
+
+void HuffmanEncoder::Reset() {
+  huf_ctable_.reset();
+  table_max_symbol_ = 0;
+}
+
+bool HuffmanEncoder::Encode(std::string_view data, uint8_t* dest, uint32_t* dest_size,
+                            std::string* error_msg) const {
+  DCHECK(huf_ctable_);
+
+  size_t res =
+      HUF_compress1X_usingCTable(dest, *dest_size, data.data(), data.size(), huf_ctable_.get(), 0);
```

Comment:
Leaving it here for info, there are:

```
/** HUF_selectDecoder() :
 *  Tells which decoder is likely to decode faster,
 *  based on a set of pre-computed metrics.
 * @return : 0==HUF_decompress4X1, 1==HUF_decompress4X2 .
 *  Assumption : 0 < dstSize <= 128 KB */
U32 HUF_selectDecoder (size_t dstSize, size_t cSrcSize);
```

Same variants exists for compression I think

Reply:
yeah i know, this can be further improvement. right now I want it relatively simple.